src/field.ts

```diff
@@ -125,7 +126,7 @@ export function isMap(fieldDescriptor: descriptor.FieldDescriptorProto) {
  * @param {descriptor.FieldDescriptorProto} fieldDescriptor 
  */
 export function isOneOf(fieldDescriptor: descriptor.FieldDescriptorProto) {
-    return typeof fieldDescriptor.oneof_index == "number";
+    return typeof pb.Message.getField(fieldDescriptor, 9) == "number"; // 9 means oneof_index
```

Comment:
I don't think this is right.

Reply:
oh, I guess you did this to circumvent some logic in the pgts generated code.  but even in that case, it shouldn't be necessary since the [proto](https://github.com/protocolbuffers/protobuf/blob/a612a75b689e12a64979d8b901d9eb503aaf8700/src/google/protobuf/descriptor.proto#L206) is in `proto2` and optional.

Reply:
This is a really complex issue.

According to the Google's documentation for proto2 (https://developers.google.com/protocol-buffers/docs/proto#optional), optionals in proto2 still have default values, even when they are not explicitly specified (oneof_index happens to be an optional with no default value specified in descriptor.proto):
_"If the default value is not specified for an optional element, a type-specific default value is used instead: for strings, the default value is the empty string. For bytes, the default value is the empty byte string. For bools, the default value is false. **For numeric types, the default value is zero**. For enums, the default value is the first value listed in the enum's type definition."_
So I guess (according to the documentation) I actually should be using `getField()` to circumvent the getter in the generated code, because the the getter would return 0, and thus the value returned by the `isOneOf()` function would always be `true`.
Also, I remember that I put those ugly `getField()` in src/descriptor.ts and src/field.ts specifically to fix the problems resulting from the generated getters returning default values.

I hope my explanation is clear and simple so far, but unfortunately there is a twist - something I discovered a few minutes ago when I played with tests to make sure that getters for proto2 optionals really return default value even when default is not explicitly specified in the proto file:
When you use a getter on `FieldDescriptorProto` class object, you really get the default value of `0` for oneof_index as expected; but if you use the `toObject()` method and then get the value of oneof_index field of the returned object, you get `undefined`.  This is evident from the generated compiler/descriptor.ts code:
```
export class FieldDescriptorProto extends pb_1.Message {
    ...
    get oneof_index() {
        return pb_1.Message.getFieldWithDefault(this, 9, 0) as number;
    }
    ...
    toObject() {
        const data: {
            ...
            oneof_index?: number;
            ...
        } = {
            ...
            oneof_index: pb_1.Message.getField(this, 9) != null ? this.oneof_index : undefined,
            ...
        };
        return data;
    }
```
Admittedly, that surprising behaviour of `toObject()` method for classes generated for proto2 messages could be taken advantage of here - to avoid the `getField()` call.  But my feeling is that it is the `toObject()` method that needs to be corrected, while the ugly `getField()` call should stay until we have `has_oneof_index` property - which we hope to add in another pull request.

To sumarize:
* I think the `getFied() call` is necessary here for now, until we have the _has_field_name_ feature.
* `toObject()` method generated for proto2 messages containing optional fields without default values specified in the proto file is buggy and needs to be fixed, because right now it returns an object without the implicit defaults, which is not even consistent with what the field getters of the class generated for the proto2 message return.

@thesayyn, @Yurzel, do you agree?

Reply:
> the generated code, because the the getter would return 0, and thus the value returned by the `isOneOf()` function would always be `true`.

this right here where I think we are missing something;  so the default value (i am specifically talking about proto2) should be in play only if a field has an `optional` prefix and has no `[default = ]` option specified.  otherwise, it is required and the `getField` call should be used instead of the `getFieldWithDefault`.

Reply:
Sorry, if I am being dumb. I just don't want to miss anything.

Reply:
> this right here where I think we are missing something; so the default value (i am specifically talking about proto2) should be in play only if a field has an `optional` prefix and has no `[default = ]` option specified. otherwise, it is required and the `getField` call should be used instead of the `getFieldWithDefault`.

I do not feel particularly clever myself about protocol buffers.  The two versions (proto2, proto3) that must be supported are already enough to give me a headache.
Anyway, I do not think that adding `[default =]` makes a field required as you implied.  My understanding is that `optional` prefix simply means that the field may not appear in a valid message (I am thinking about the serialized form of the message here).  Furthermore, I think both with `[default = some_value]` and without it the field has a default value.  Omitting `[default = some_value]` has exactly the same effect for optional numeric fields as specifying `[default = 0]`.

In other words, to use a slightly modified example from Google documentation, I think that
```
optional int32 result_per_page = 3 [default = 0];
```
works exactly the same as:
```
optional int32 result_per_page = 3;
```
In both cases the field is optional and has a default value of zero.  Obviously, a user of a class generated for a protobuf message containing such a default field will see that `object_of_that_class.results_per_page == 0` even, if `object_of_that_class` is a newly-created object, on which no fields have been explicitly set yet.  But when the object is serialized, the serialization code will notice that no value has been explicitly set for `object_of_that_class.results_per_page` and will omit that field during serialization.  Per my understanding, in proto2 explicitly assigning a value to a field, even if you set it to the default value (i.e. `object_of_that_class.results_per_page = 0`) will prevent serialization code from omitting that field (serialization works differently for proto2 and proto3 here).
To put it differently, if a numeric field is optional in proto2, a getter for that field (in the generated class) should always return some value for that field (if the field is not present, it simply returns the default value, which might be specified explicitly via `[default =]` or implicitly be 0).  The optionality mostly makes a difference during serialization.
I might be wrong so I am asking that anybody who has a deeper understanding of proto2 corrects me.

Reply:
> Anyway, I do not think that adding `[default =]` makes a field required as you implied.

I wasn't trying to say this.  I know that it doesn't. I just wonder what happens if I don't mark a field optional and not set it in proto2.

Reply:
I think it should return undefined instead of the default since it is not marked as `optional`. Let's add a test case for this to assert that it doesn't slip away.

Reply:
and you can keep the getField call until we have absence fields.

Reply:
Sounds right to me. Also I made test protofiles test/_/default/ **default_proto2** and **default_proto3**. You can run them with protoc (3.20 because in 3.21 javascript was accidentally? removed) and then compare getters. You can do the same with field presence later with test/\_/presence/ **presence_proto2** and **presence_proto3**.

As far as I remember **optional** has no effect on getters or defaults. The only effect it has on is whenever the field presence is tracked or not. At least in proto3.

Reply:
> > Anyway, I do not think that adding `[default =]` makes a field required as you implied.
> 
> I wasn't trying to say this. I know that it doesn't. I just wonder what happens if I don't mark a field optional and not set it in proto2.

Ah, that is a question that is hard to answer definitely. If you don't mark a field as optional in proto2, then I assume you must mark it as required (in proto2 you have to mark every numeric [scalar] field as optional, or required, or repeated). When a required field is not set to any value, then the proto2 message is definitely invalid. But how that invalid message will be handled is apparently implementation-defined.  Java users might get `com.google.protobuf.InvalidProtocolBufferException` exception with a message saying _"Message was
missing required fields"_ when unserializing such an invalid message.  But the javascript implementation from Google (I tried v3.20 as suggested by @Yurzel) happily serializes and unserializes such invalid messages (missing a required field) without complains.  I suppose the conclusion is that a proto2 message missing a required field is invalid, but the consequences of the message being invalid are not clearly specified - kind of like 'undefined behaviour' for C language compilers.

> I think it should return undefined instead of the default since it is not marked as optional. Let's add a test case for this to assert that it doesn't slip away.

I would agree with that, but the field in question (oneof_index, id = 9 in `message FieldDescriptorProto`) apparently is marked as optional:
https://github.com/protocolbuffers/protobuf/blob/a612a75b689e12a64979d8b901d9eb503aaf8700/src/google/protobuf/descriptor.proto#L206
Can I keep the `getField()` trick here as well?  Or am I somehow looking at the wrong .proto file?

Reply:
Missing a required field should be indeed invalid. C# implementation doesn't check it either. I think a better way to look at it is the `required`, `optional` and `repeated` stands for the serialized data or "what will go over the wire".

proto2
- optional: isn't needed to be set
  - optimalization: don't send anything in case of undefined
- required: must be set
- repeated: array

proto3
- optional: isn't needed to be set
  - optimalization: don't send anything in case of undefined
- required: this keyword doesn't exist, because you don't have to set the value, it will be the default value of the type
  - optimalization: don't send default value, because the other side knows what the default is
- repeated: array

The `[default =]` just says what the getter will return in case of undefined, no matter what. There is no reason to have it on `required` field but I guess it doesn't break anything.